kerastuner/engine/conditions.py

```diff
@@ -0,0 +1,150 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"HyperParameters logic."
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import abc
+import six
+
+from ..protos import kerastuner_pb2
+
+
+@six.add_metaclass(abc.ABCMeta)
+class Condition(object):
+    """Abstract condition for a conditional hyperparameter.
+
+    Subclasses of this object can be passed to a `HyperParameter` to
+    specify that this condition must be met in order for that hyperparameter
+    to be considered active for the `Trial`.
+
+    Example:
+
+    ```
+    condition = kt.conditions.OneOf('a', 'dnn')
+
+    a = Choice('model', ['linear', 'dnn'])
+    b = Int('num_layers', 5, 10, conditions=[condition])
+    ```
+    """
+
+    @abc.abstractmethod
+    def is_active(self, values):
+        """Whether this condition should be considered active.
+
+        Determines whether this condition is true for the current `Trial`.
+
+        # Arguments:
+            values: Dict. The active values for this `Trial`. Keys are the
+               names of the hyperparameters.
+
+        # Returns:
+            bool.
+        """
+        raise NotImplementedError('Must be implemented in subclasses.')
+
+    @abc.abstractmethod
+    def __eq__(self, other):
+        raise NotImplementedError('Must be implemented in subclasses.')
+
+    @abc.abstractmethod
+    def get_config(self):
+        raise NotImplementedError('Must be implemented in subclasses.')
+
+    @classmethod
+    def from_config(cls, config):
+        return cls(**config)
+
+    @classmethod
+    def from_proto(self, proto):
+        kind = proto.WhichOneof('kind')
+        if kind == 'parent':
+            parent = getattr(proto, kind)
+            name = parent.name
+            values = parent.values
+            values = [getattr(v, v.WhichOneof('kind')) for v in values]
+            return Parent(name=name, values=values)
+        raise ValueError('Unrecognized condition of type: {}'.format(kind))
+
+
+class Parent(Condition):
+    """Condition that checks a value is equal to one of a list of values.
+
+    This object can be passed to a `HyperParameter` to specify that this
+    condition must be met in order for that hyperparameter to be considered
+    active for the `Trial`.
+
+    Example:
+
+    ```
+    a = Choice('model', ['linear', 'dnn'])
+    b = Int('num_layers', 5, 10, conditions=[kt.conditions.Parent('a', ['dnn'])])
```

Comment:
Isn't it a problem that the parent is specified by name only, and not by the full path to the parent?

Imagine the following conditional tree:
parent(name='model', values=['dnn', 'linear'])
child(name='lr', values=[0.01, 0.1], conditions=[kt.conditions,Parent('model',['dnn'])])
child(name='num_layers', 1, 5, conditions=[kt.conditions,Parent('model',['dnn'])])
child(name='lr', values=[0.01, 0.2], conditions=[kt.conditions,Parent('model',['linear'])])
child(name='lr_func', values=['exp', 'beta'], conditions=[kt.conditions,Parent('lr',[0.2])])

In this case, there are two 'lr' parameters, but active for different parent values.
But 'lr_func' specifies a parent 'lr'. How do we know which 'lr' we mean here?

Reply:
Note that the `Conditions` API is not public-facing, it's accessed via `HyperParameters.conditional_scope` and `parent_name` and `parent_values` arguments in `HyperParameters.{Choice | Int | etc.}`

I think we're locked into using `parent_name` and `parent_values` to express this, since this is in the public API for `HyperParameters` via `hps.conditional_scope` and via the `parent_name` and `parent_values` arguments for `hps.Choice`, `hps.Int`, etc

There is some possibility of ambiguity when going from `HyperParameters` to the internal representation, however we only need the other, one-way conversion for the internal Tuner (internal to HyperParameters) and this is not ambiguous, since one `Parent` condition will be added for each node in the internal tree

If we decide to add two-way conversion to the internal tuner, we can error out if an ambiguity is discovered

Mostly I think this won't be an issue though, since conditional hyperparameters are created one of 2 ways in Keras Tuner:

(1) With conditional scopes

```python
hps = kt.HyperParameters()
hps.Choice('model', ['dnn', 'linear'])
with hps.conditional_scope('model', 'dnn'):
  hps.Int('layers', 5, 10)
  with hps.conditional_scope('layers', 10):
    hps.Int('units_10', 10, 100, step=10)
```

In this case, "units_10" will have two conditions: model==dnn and layers==10. Therefore there will be no ambiguity

(2) Directly via `parent_name` and `parent_values`

```python
hps = kt.HyperParameters()
hps.Choice('model', ['dnn', 'linear'])
hps.Int('layers', 5, 10, parent_name='dnn', parent_values='linear')
hps.Int('units_10', 10, 100, step=10, parent_name='layers', parent_values=10)
```

In this case, for Keras Tuner purposes it is still ok, because Keras Tuner only needs to check if these exists some param named 'layers' that has a value of 10. But if we end up needing to convert HyperParameters to internal representation, we should raise an error saying this is ambiguous